plutarch-benchmark/src/Plutarch/Benchmark.hs

```diff
@@ -49,29 +59,120 @@ import Data.Maybe (fromJust)
 import GHC.Generics (Generic)
 import GHC.Stack (HasCallStack)
 import Plutarch (ClosedTerm, compile, printTerm)
+import Plutarch.Api.V1 (
+  PScriptContext,
+  type PMintingPolicy,
+  type PStakeValidator,
+  type PValidator,
+ )
+import Plutarch.Prelude
+import Plutarch.Unsafe (punsafeCoerce)
 import Plutus.V1.Ledger.Api (
   ExBudget (ExBudget),
   ExCPU (ExCPU),
   ExMemory (ExMemory),
   Script,
  )
 import qualified Plutus.V1.Ledger.Api as Plutus
+import qualified Plutus.V1.Ledger.Scripts as Plutus
 
 --------------------------------------------------------------------------------
 
 -- | Benchmark the given script
 benchmarkScript :: String -> Script -> NamedBenchmark
-benchmarkScript name = NamedBenchmark . (name,) . benchmarkScript'
+benchmarkScript name =
+  NamedBenchmark . (name,) . benchmarkScript' Nothing
+
+{- | Benchmark the given script,
+  with a provided un-applied version for sizing
+-}
+benchmarkScriptUnapplied :: String -> Script -> Script -> NamedBenchmark
+benchmarkScriptUnapplied name unApplied =
+  NamedBenchmark . (name,) . benchmarkScript' (Just unApplied)
+
+{- |
+  Benchmark the given script, using a function to apply arguments.
+-}
+benchmarkScriptWithApply :: String -> Script -> (Script -> Script) -> NamedBenchmark
+benchmarkScriptWithApply name unApplied applyArgs =
+  benchmarkScriptUnapplied name unApplied $ applyArgs unApplied
+
+-- | Benchmark a Validator with the provided arguments
+benchmarkValidator ::
+  ( Plutus.ToData d
+  , Plutus.ToData r
+  ) =>
+  String ->
+  Plutus.Validator ->
+  d ->
+  r ->
+  Plutus.ScriptContext ->
+  NamedBenchmark
+benchmarkValidator name (Plutus.Validator unApplied) datum redeemer ctx =
+  benchmarkScriptUnapplied name unApplied $
+    Plutus.applyArguments
+      unApplied
+      [ Plutus.toData datum
+      , Plutus.toData redeemer
+      , Plutus.toData ctx
+      ]
+
+-- | Benchmark a MintingPolicy with the provided arguments
+benchmarkMintingPolicy ::
```

Comment:
I'm honestly not in favor of adding functions specific to each "type" of script. Rather than using `Plutus.Validator`, etc., these should just take the term, and the user should just specify what the type of the term is, e.g. `PValidator`.

Reply:
removed them, benchUnapplied should suffice I think.